Mathlib/RingTheory/TwoSidedIdeal.lean

```diff
@@ -0,0 +1,227 @@
+
+instance : SMul ℤ I where
+  smul n x := ⟨n • x.1, I.zsmul_mem n x.2⟩
+
+instance : AddCommGroup I :=
+  Function.Injective.addCommGroup _ Subtype.coe_injective
+    rfl (fun _ _ ↦ rfl) (fun _ ↦ rfl) (fun _ _ ↦ rfl) (fun _ _ ↦ rfl) (fun _ _ ↦ rfl)
+
+/-- The coercion into the ring as a `AddMonoidHom` -/
+@[simp]
+def coeAddMonoidHom : I →+ R where
+  toFun := (↑)
+  map_zero' := rfl
+  map_add' _ _ := rfl
+
+end NonUnitalNonAssocRing
+
+section ring
+
+variable {R : Type*} [Ring R] (I : RingCon R)
+
+instance : SMul R I where smul r x := ⟨r • x.1, I.mul_mem_left _ _ x.2⟩
+
+instance : SMul Rᵐᵒᵖ I where smul r x := ⟨r • x.1, I.mul_mem_right _ _ x.2⟩
+
+instance : Module R I :=
+  Function.Injective.module _ (coeAddMonoidHom I) Subtype.coe_injective fun _ _ => rfl
+
+instance : Module Rᵐᵒᵖ I :=
+  Function.Injective.module _ (coeAddMonoidHom I) Subtype.coe_injective fun _ _ => rfl
+
+instance : SMulCommClass R Rᵐᵒᵖ I where
+  smul_comm r s x := Subtype.ext <| smul_comm r s x.1
+
+/--
+For any `I : RingCon R`, when we view it as an ideal, `I.subtype` is the injective `R`-linear map
+`I → R`.
+-/
+@[simps]
+def subtype : I →ₗ[R] R where
+  toFun x := x.1
+  map_add' _ _ := rfl
+  map_smul' _ _ := rfl
+
+/--
+For any `RingCon R`, when we view it as an ideal in `Rᵒᵖ`, `subtype` is the injective `Rᵐᵒᵖ`-linear
+map `I → Rᵐᵒᵖ`.
+-/
+@[simps]
+def subtypeMop : I →ₗ[Rᵐᵒᵖ] Rᵐᵒᵖ where
+  toFun x := MulOpposite.op x.1
+  map_add' _ _ := rfl
+  map_smul' _ _ := rfl
+
+end ring
+
+section operations
+
+variable {R : Type*} [Ring R] (I : RingCon R)
+
+/--
+Every two-sided ideal is also a left ideal.
+-/
+def toIdeal : Ideal R where
+  carrier := I
+  add_mem' := I.add_mem
+  zero_mem' := I.zero_mem
+  smul_mem' := I.mul_mem_left
+
+lemma mem_toIdeal {x} : x ∈ I.toIdeal ↔ x ∈ I := Iff.rfl
+
+/--
+Every two-sided ideal is also a right ideal.
+-/
+def toIdealMop : Ideal Rᵐᵒᵖ := LinearMap.range I.subtypeMop
+
+lemma mem_toIdealMop {x} : x ∈ I.toIdealMop ↔ x.unop ∈ I := by
+  simp only [toIdealMop, LinearMap.mem_range, subtypeMop_apply, Subtype.exists, exists_prop]
+  constructor
+  · rintro ⟨a, ha, rfl⟩; simpa
+  · intro h; exact ⟨x.unop, h, rfl⟩
+
+lemma mem_toIdealMop' {x} : (MulOpposite.op x) ∈ I.toIdealMop ↔ x ∈ I := by
+  rw [mem_toIdealMop]; rfl
+
+/--
+The smallest two-sided-ideal contain a set.
+-/
+def span (s : Set R) : RingCon R := ringConGen (fun a b ↦ a - b ∈ s)
+
+lemma subset_span {s : Set R} : s ⊆ (span s : Set R) := by
+  intro x hx
+  rw [SetLike.mem_coe, mem_iff]
+  exact RingConGen.Rel.of _ _ (by simpa using hx)
+
+lemma mem_span_iff {s : Set R} {x} :
+    x ∈ span s ↔ ∀ (I : RingCon R), s ⊆ I → x ∈ I := by
+  refine ⟨?_, fun h => h _ subset_span⟩
+  delta span
+  rw [ringConGen_eq]
+  intro h I hI
+  refine sInf_le (α := RingCon R) ?_ h
+  intro x y hxy
+  specialize hI hxy
+  rwa [SetLike.mem_coe, ← rel_iff] at hI
+
+/--
+Every left ideal generates a two sided ideal.
+-/
+def fromIdeal (I : Ideal R) : RingCon R := span I
```

Comment:
This exists already, it's `Ideal.Quotient.ringCon`

Reply:
`Ideal.Quotient.ringCon` needs the commutativity assumption. Not sure if it really need it or the commutativity assumption was added by accident

Reply:
```lean
protected def ringCon (I : Ideal R) : RingCon R :=
  { QuotientAddGroup.con I.toAddSubgroup with
    mul' := fun {a₁ b₁ a₂ b₂} h₁ h₂ => by
      rw [Submodule.quotientRel_r_def] at h₁ h₂ ⊢
      have F := I.add_mem (I.mul_mem_left a₂ h₁) (I.mul_mem_right b₁ h₂)
      have : a₁ * a₂ - b₁ * b₂ = a₂ * (a₁ - b₁) + (a₂ - b₂) * b₁ := by
        rw [mul_sub, sub_mul, sub_add_sub_cancel, mul_comm, mul_comm b₁]
      rwa [← this] at F }
```
seems the proof actually used `mul_comm`

Reply:
If I'm not mistaken, `Ideal.Quoteint.ringCon` will ensure that `I : Ideal R` when seen as a two-sided-ideal is exactly the same as `I` (because `R` is commutative). But `fromIdeal I` might be bigger than `I` because we are multiplying things on the right side of the left idea I?

Reply:
Can we also have a bundling as an order isomorphism when `R` is commutative, please?